org.jacoco.core/src/org/jacoco/core/tools/ExecDumpClient.java

```diff
@@ -123,7 +123,10 @@ public ExecFileLoader dump(final InetAddress address, final int port)
 					.setExecutionDataVisitor(loader.getExecutionDataStore());
 
 			remoteWriter.visitDumpCommand(dump, reset);
-			remoteReader.read();
+
+			if (!remoteReader.read()) {
+				throw new IOException("Socket closed unexpectedly.");
```

Comment:
@marchof understanding why real agent can close connection prior to answering dump does not seem so trivial, so IMO this definitely deserves separate PR. No need to revert change here - just create separate, once its merge into master branch, merge of master into this one will result in desired effect.